clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -1254,28 +1274,68 @@ scheduler_tmpl_t<RecordType, ReaderType>::read_switch_sequences()
 typename scheduler_tmpl_t<RecordType, ReaderType>::scheduler_status_t
-scheduler_tmpl_t<RecordType, ReaderType>::get_initial_timestamps()
-{
+scheduler_tmpl_t<RecordType, ReaderType>::get_initial_input_content(
+    bool gather_timestamps)
+{
+    // For every mode, read ahead until we see a filetype record so the user can
+    // examine it prior to retrieving any records.
+    VPRINT(this, 1, "Reading headers from inputs to find filetypes%s\n",
+           gather_timestamps ? " and timestamps" : "");
+    assert(options_.read_inputs_in_init);
     // Read ahead in each input until we find a timestamp record.
     // Queue up any skipped records to ensure we present them to the
     // output stream(s).
     for (size_t i = 0; i < inputs_.size(); ++i) {
         input_info_t &input = inputs_[i];
-        if (input.next_timestamp <= 0) {
+        bool found_filetype = false;
+        bool found_timestamp = !gather_timestamps || input.next_timestamp > 0;
+        if (!found_filetype || !found_timestamp) {
             for (const auto &record : input.queue) {
+                trace_marker_type_t marker_type;
+                uintptr_t marker_value;
+                if (record_type_is_marker(record, marker_type, marker_value) &&
+                    marker_type == TRACE_MARKER_TYPE_FILETYPE) {
+                    found_filetype = true;
+                    VPRINT(this, 2, "Input %zu filetype %zu\n", i, marker_value);
+                }
                 if (record_type_is_timestamp(record, input.next_timestamp))
+                    found_timestamp = true;
+                if (found_filetype && found_timestamp)
                     break;
             }
         }
-        if (input.next_timestamp <= 0) {
+        if (input.next_timestamp > 0)
+            found_timestamp = true;
+        if (!found_filetype || !found_timestamp) {
             if (input.needs_init) {
                 input.reader->init();
                 input.needs_init = false;
             }
             while (*input.reader != *input.reader_end) {
                 RecordType record = **input.reader;
+                trace_marker_type_t marker_type;
+                uintptr_t marker_value;
+                if (record_type_is_marker(record, marker_type, marker_value) &&
+                    marker_type == TRACE_MARKER_TYPE_FILETYPE) {
+                    found_filetype = true;
+                    VPRINT(this, 2, "Input %zu filetype %zu\n", i, marker_value);
+                }
                 if (record_type_is_timestamp(record, input.next_timestamp))
+                    found_timestamp = true;
+                if (found_filetype && found_timestamp)
                     break;
+                // Don't go too far if only looking for filetype, to avoid reaching
+                // the first instruction, which messes up some unit tests that don't
+                // bother to include a filetype.  Just exit with a 0 filetype.
+                if (!found_filetype &&
```

Comment:
Wouldn't it be better to modify those unit tests to include a filetype (if possible)? These special-case checks make it harder to read sometimes.

Reply:
It turns out legacy traces don't have filetypes either, including the test ones checked in, and hitting the instruction can hit ordinal problems (maybe taking over the ordinals completely would help that...there is a comment on that added in this PR).  Updated the comment.